src/ripple/app/ledger/impl/LedgerMaster.cpp

```diff
@@ -1037,6 +1039,53 @@ LedgerMaster::checkAccept(std::shared_ptr<Ledger const> const& ledger)
     app_.getFeeTrack().setRemoteFee(fee);
 
     tryAdvance();
+
+    if (ledger->seq() % 256 == 0)
+    {
+        // Check if the majority of validators run a higher version rippled
+        // software. If so print a warning.
+        //
+        // Once the HardenedValidations amendment is enabled, validators include
+        // their rippled software version in the validation messages of every
+        // (flag - 1) ledger. We wait for one ledger time before checking the
+        // version information to accumulate more validation messages.
+
+        auto const& parentHash = ledger->info().parentHash;
+        auto const parentLedger = getLedgerByHash(parentHash);
+        if (parentLedger &&
+            parentLedger->rules().enabled(featureHardenedValidations))
+        {
+            // To throttle the warning messages, instead of printing a warning
+            // every flag ledger, we print every week.
+            auto currentTime = app_.timeKeeper().now();
+            if (currentTime - upgradeWarningPrevTime_ >= weeks{1})
+            {
+                auto const vals =
+                    app_.getValidations().getTrustedForLedger(parentHash);
+                auto higherVersionCount = std::count_if(
+                    vals.begin(), vals.end(), [](auto const& v) -> bool {
+                        if (v->isFieldPresent(sfServerVersion))
+                            return BuildInfo::isNewerVersion(
+                                v->getFieldU64(sfServerVersion));
+                        else
+                            return false;
+                    });
+                // We set the threshold of majority to be 60% of the UNL
```

Comment:
We do this kind of thing often enough that a function with well-defined and consistent semantics would be useful:

```
// This version rounds up. For example:
// (6 / 14) = 42.8571% and this function returns 43
// (29, 33) = 87.8788% and this function returns 88
// (1, 64) = 1.5625% and this function returns 2
int
calculatePercent(int count, int total)
{
    assert (total != 0);
    return ((std::min(count, total) * 100) + total - 1) / total;
}
```

This check then becomes:

```
constexpr int minimumPercentage = 60;

if (calculatePercent(higherVersionCount, app_.validators().getQuorumKeys().second.size()) >= minimumPercentage)
{
    ...
}
```

Reply:
> `return ((std::min(count, total) * 100) + total - 1) / total;`

can also be written as
`return ((std::min(count, total) * 100) - 1) / total + 1;`
The algebra is pretty obvious, but I find the "minus one, plus one" a little easier to read, because I'm looking at constants instead of more variable names.

Reply:
Thanks. I added the calculatePercent function and unit tests. I used `std::size_t` since I don't think we need negative numbers. 
`std::size_t
calculatePercent(std::size_t count, std::size_t total);`

`return ((std::min(count, total) * 100) - 1) / total + 1;` works for int. Since `std::size_t` is used, it is not used because of possible underflow.